File: agents/typescript/fw-agent/src/poller.test.ts
// Copyright 2025 Ralph Lemke
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { AgentPoller, Handler } from "./poller";
import { defaultConfig } from "./config";
import {
  CollectionTasks,
  CollectionSteps,
  CollectionServers,
  CollectionStepLogs,
  TaskStatePending,
  TaskStateRunning,
  TaskStateCompleted,
  ResumeTaskName,
  StepLogLevelInfo,
  StepLogLevelWarning,
  StepLogLevelError,
  StepLogLevelSuccess,
  StepLogSourceFramework,
  StepLogSourceHandler,
} from "./protocol";
import { inferTypeHint } from "./models";

describe("AgentPoller", () => {
  it("should create with default config", () => {
    const config = defaultConfig();
    const poller = new AgentPoller(config);

    expect(poller).toBeDefined();
    expect(poller.registeredHandlers()).toHaveLength(0);
  });

  it("should register handlers", () => {
    const config = defaultConfig();
    const poller = new AgentPoller(config);

    const handler: Handler = async (params) => {
      return { result: "ok" };
    };

    poller.register("ns.TestFacet", handler);

    const handlers = poller.registeredHandlers();
    expect(handlers).toHaveLength(1);
    expect(handlers[0]).toBe("ns.TestFacet");
  });

  it("should register multiple handlers", () => {
    const config = defaultConfig();
    const poller = new AgentPoller(config);

    poller.register("ns.FacetA", async () => ({}));
    poller.register("ns.FacetB", async () => ({}));
    poller.register("FacetC", async () => ({}));

    const handlers = poller.registeredHandlers();
    expect(handlers).toHaveLength(3);
    expect(handlers).toContain("ns.FacetA");
    expect(handlers).toContain("ns.FacetB");
    expect(handlers).toContain("FacetC");
  });

  it("should inject _facet_name into params", () => {
    const poller = new AgentPoller(defaultConfig());
    poller.register("ns.TestFacet", async (params) => {
      expect(params["_facet_name"]).toBe("ns.TestFacet");
      return {};
    });
    // Verify the handler is registered (integration test would cover full injection)
    expect(poller.registeredHandlers()).toContain("ns.TestFacet");
  });

  it("should have null metadataProvider by default", () => {
    const poller = new AgentPoller(defaultConfig());
    expect(poller.metadataProvider).toBeNull();
  });

  it("should use metadataProvider when set", () => {
    const poller = new AgentPoller(defaultConfig());
    poller.metadataProvider = (facetName: string) => {
      if (facetName === "ns.TestFacet") {
        return { description: "test handler" };
      }
      return undefined;
    };
    const meta = poller.metadataProvider("ns.TestFacet");
    expect(meta).toEqual({ description: "test handler" });
    expect(poller.metadataProvider("ns.Other")).toBeUndefined();
  });

  it("should inject _update_step callback into params", () => {
    const poller = new AgentPoller(defaultConfig());
    poller.register("ns.StreamFacet", async (params) => {
      expect(params["_update_step"]).toBeDefined();
      return {};
    });
    expect(poller.registeredHandlers()).toContain("ns.StreamFacet");
  });

  it("should have _update_step as async function type", () => {
    // Verify the callback signature pattern
    const updateStep = async (partial: Record<string, unknown>) => {
      // Simulated partial update
    };
    expect(typeof updateStep).toBe("function");
  });

  it("should support partial updates via _update_step", () => {
    // Verify the pattern: callback receives partial returns
    const updates: Record<string, unknown>[] = [];
    const updateStep = async (partial: Record<string, unknown>) => {
      updates.push(partial);
    };
    updateStep({ progress: 50 });
    updateStep({ progress: 100, result: "done" });
    expect(updates).toHaveLength(2);
  });
});

describe("defaultConfig", () => {
  it("should return expected defaults", () => {
    const config = defaultConfig();

    expect(config.serviceName).toBe("fw-agent");
    expect(config.serverGroup).toBe("default");
    expect(config.taskList).toBe("default");
    expect(config.maxConcurrent).toBe(5);
    expect(config.pollIntervalMs).toBe(2000);
    expect(config.heartbeatIntervalMs).toBe(10000);
    expect(config.mongoUrl).toBe("mongodb://localhost:27017");
    expect(config.database).toBe("afl");
  });
});

describe("Protocol constants", () => {
  it("should have correct collection names", () => {
    expect(CollectionTasks).toBe("tasks");
    expect(CollectionSteps).toBe("steps");
    expect(CollectionServers).toBe("servers");
  });

  it("should have correct task states", () => {
    expect(TaskStatePending).toBe("pending");
    expect(TaskStateRunning).toBe("running");
    expect(TaskStateCompleted).toBe("completed");
  });

  it("should have correct protocol task name", () => {
    expect(ResumeTaskName).toBe("fw:resume");
  });
});

describe("Step log constants", () => {
  it("should have correct step log levels", () => {
    expect(StepLogLevelInfo).toBe("info");
    expect(StepLogLevelWarning).toBe("warning");
    expect(StepLogLevelError).toBe("error");
    expect(StepLogLevelSuccess).toBe("success");
  });

  it("should have correct step log sources", () => {
    expect(StepLogSourceFramework).toBe("framework");
    expect(StepLogSourceHandler).toBe("handler");
  });

  it("should have correct step_logs collection", () => {
    expect(CollectionStepLogs).toBe("step_logs");
  });
});

describe("inferTypeHint", () => {
  it("should infer Boolean", () => {
    expect(inferTypeHint(true)).toBe("Boolean");
    expect(inferTypeHint(false)).toBe("Boolean");
  });

  it("should infer Long for integers", () => {
    expect(inferTypeHint(42)).toBe("Long");
    expect(inferTypeHint(0)).toBe("Long");
    expect(inferTypeHint(-100)).toBe("Long");
  });

  it("should infer Double for floats", () => {
    expect(inferTypeHint(3.14)).toBe("Double");
    expect(inferTypeHint(0.5)).toBe("Double");
  });

  it("should infer String", () => {
    expect(inferTypeHint("hello")).toBe("String");
    expect(inferTypeHint("")).toBe("String");
  });

  it("should infer List for arrays", () => {
    expect(inferTypeHint([1, 2, 3])).toBe("List");
    expect(inferTypeHint([])).toBe("List");
  });

  it("should infer Map for objects", () => {
    expect(inferTypeHint({ a: 1 })).toBe("Map");
    expect(inferTypeHint({})).toBe("Map");
  });

  it("should infer Any for null/undefined", () => {
    expect(inferTypeHint(null)).toBe("Any");
    expect(inferTypeHint(undefined)).toBe("Any");
  });
});

File: agents/go/fw-agent/poller_test.go
// Copyright 2025 Ralph Lemke
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package fwagent

import (
	"testing"
)

func TestNewAgentPoller(t *testing.T) {
	cfg := DefaultConfig()
	poller := NewAgentPoller(cfg)

	if poller == nil {
		t.Fatal("NewAgentPoller returned nil")
	}

	if poller.serverID == "" {
		t.Error("serverID should not be empty")
	}

	if poller.handlers == nil {
		t.Error("handlers map should be initialized")
	}
}

func TestRegisterHandler(t *testing.T) {
	cfg := DefaultConfig()
	poller := NewAgentPoller(cfg)

	handler := func(params map[string]interface{}) (map[string]interface{}, error) {
		return map[string]interface{}{"result": "ok"}, nil
	}

	poller.Register("ns.TestFacet", handler)

	handlers := poller.RegisteredHandlers()
	if len(handlers) != 1 {
		t.Errorf("Expected 1 handler, got %d", len(handlers))
	}

	if handlers[0] != "ns.TestFacet" {
		t.Errorf("Expected handler name 'ns.TestFacet', got '%s'", handlers[0])
	}
}

func TestFindHandler(t *testing.T) {
	cfg := DefaultConfig()
	poller := NewAgentPoller(cfg)

	handler := func(params map[string]interface{}) (map[string]interface{}, error) {
		return map[string]interface{}{"result": "ok"}, nil
	}

	// Register with short name
	poller.Register("TestFacet", handler)

	// Test exact match
	h := poller.findHandler("TestFacet")
	if h == nil {
		t.Error("Should find handler by exact match")
	}

	// Test short name fallback
	h = poller.findHandler("ns.TestFacet")
	if h == nil {
		t.Error("Should find handler by short name fallback")
	}

	// Test not found
	h = poller.findHandler("ns.Unknown")
	if h != nil {
		t.Error("Should return nil for unknown handler")
	}
}

func TestDefaultConfig(t *testing.T) {
	cfg := DefaultConfig()

	if cfg.ServiceName != "fw-agent" {
		t.Errorf("Expected ServiceName 'fw-agent', got '%s'", cfg.ServiceName)
	}

	if cfg.ServerGroup != "default" {
		t.Errorf("Expected ServerGroup 'default', got '%s'", cfg.ServerGroup)
	}

	if cfg.TaskList != "default" {
		t.Errorf("Expected TaskList 'default', got '%s'", cfg.TaskList)
	}

	if cfg.MaxConcurrent != 5 {
		t.Errorf("Expected MaxConcurrent 5, got %d", cfg.MaxConcurrent)
	}

	if cfg.MongoURL != "mongodb://localhost:27017" {
		t.Errorf("Expected MongoURL 'mongodb://localhost:27017', got '%s'", cfg.MongoURL)
	}

	if cfg.Database != "afl" {
		t.Errorf("Expected Database 'afl', got '%s'", cfg.Database)
	}
}

func TestProtocolConstants(t *testing.T) {
	// Verify constants match protocol
	if CollectionTasks != "tasks" {
		t.Errorf("CollectionTasks should be 'tasks', got '%s'", CollectionTasks)
	}

	if CollectionSteps != "steps" {
		t.Errorf("CollectionSteps should be 'steps', got '%s'", CollectionSteps)
	}

	if CollectionServers != "servers" {
		t.Errorf("CollectionServers should be 'servers', got '%s'", CollectionServers)
	}

	if TaskStatePending != "pending" {
		t.Errorf("TaskStatePending should be 'pending', got '%s'", TaskStatePending)
	}

	if TaskStateRunning != "running" {
		t.Errorf("TaskStateRunning should be 'running', got '%s'", TaskStateRunning)
	}

	if TaskStateCompleted != "completed" {
		t.Errorf("TaskStateCompleted should be 'completed', got '%s'", TaskStateCompleted)
	}

	if ResumeTaskName != "fw:resume" {
		t.Errorf("ResumeTaskName should be 'afl:resume', got '%s'", ResumeTaskName)
	}
}

func TestStepLogConstants(t *testing.T) {
	// Step log levels
	if StepLogLevelInfo != "info" {
		t.Errorf("StepLogLevelInfo should be 'info', got '%s'", StepLogLevelInfo)
	}
	if StepLogLevelWarning != "warning" {
		t.Errorf("StepLogLevelWarning should be 'warning', got '%s'", StepLogLevelWarning)
	}
	if StepLogLevelError != "error" {
		t.Errorf("StepLogLevelError should be 'error', got '%s'", StepLogLevelError)
	}
	if StepLogLevelSuccess != "success" {
		t.Errorf("StepLogLevelSuccess should be 'success', got '%s'", StepLogLevelSuccess)
	}

	// Step log sources
	if StepLogSourceFramework != "framework" {
		t.Errorf("StepLogSourceFramework should be 'framework', got '%s'", StepLogSourceFramework)
	}
	if StepLogSourceHandler != "handler" {
		t.Errorf("StepLogSourceHandler should be 'handler', got '%s'", StepLogSourceHandler)
	}

	// Step logs collection
	if CollectionStepLogs != "step_logs" {
		t.Errorf("CollectionStepLogs should be 'step_logs', got '%s'", CollectionStepLogs)
	}

	// Handler registrations collection
	if CollectionHandlerRegistrations != "handler_registrations" {
		t.Errorf("CollectionHandlerRegistrations should be 'handler_registrations', got '%s'",
			CollectionHandlerRegistrations)
	}
}

func TestFacetNameInjection(t *testing.T) {
	cfg := DefaultConfig()
	poller := NewAgentPoller(cfg)

	var receivedFacetName interface{}
	handler := func(params map[string]interface{}) (map[string]interface{}, error) {
		receivedFacetName = params["_facet_name"]
		return nil, nil
	}

	poller.Register("ns.TestFacet", handler)

	// Simulate what processTask does: read params, inject, call handler
	params := map[string]interface{}{}
	params["_facet_name"] = "ns.TestFacet"
	handler(params)

	if receivedFacetName != "ns.TestFacet" {
		t.Errorf("Expected _facet_name 'ns.TestFacet', got '%v'", receivedFacetName)
	}
}

func TestHandlerMetadataInjection(t *testing.T) {
	cfg := DefaultConfig()
	poller := NewAgentPoller(cfg)

	// Set a metadata provider
	poller.metadataProvider = func(facetName string) map[string]interface{} {
		if facetName == "ns.TestFacet" {
			return map[string]interface{}{"description": "test handler"}
		}
		return nil
	}

	meta := poller.metadataProvider("ns.TestFacet")
	if meta == nil {
		t.Fatal("Expected metadata, got nil")
	}
	if meta["description"] != "test handler" {
		t.Errorf("Expected description 'test handler', got '%v'", meta["description"])
	}
}

func TestHandlerMetadataAbsent(t *testing.T) {
	cfg := DefaultConfig()
	poller := NewAgentPoller(cfg)

	// No metadata provider set — should not panic
	if poller.metadataProvider != nil {
		t.Error("metadataProvider should be nil by default")
	}
}

func TestUpdateStepCallbackInjection(t *testing.T) {
	cfg := DefaultConfig()
	poller := NewAgentPoller(cfg)

	var receivedCallback interface{}
	handler := func(params map[string]interface{}) (map[string]interface{}, error) {
		receivedCallback = params["_update_step"]
		return nil, nil
	}

	poller.Register("ns.StreamFacet", handler)

	// Simulate param injection
	params := map[string]interface{}{}
	params["_update_step"] = func(partial map[string]interface{}) {}
	handler(params)

	if receivedCallback == nil {
		t.Error("Expected _update_step callback to be injected")
	}
}

func TestUpdateStepCallbackType(t *testing.T) {
	cfg := DefaultConfig()
	poller := NewAgentPoller(cfg)

	// Verify that the callback type matches expected signature
	var cb func(map[string]interface{})
	cb = func(partial map[string]interface{}) {
		// Simulated partial update
		if partial == nil {
			t.Error("partial should not be nil")
		}
	}

	params := map[string]interface{}{"_update_step": cb}
	fn, ok := params["_update_step"].(func(map[string]interface{}))
	if !ok {
		t.Fatal("_update_step should be func(map[string]interface{})")
	}
	fn(map[string]interface{}{"progress": 50})

	_ = poller // ensure poller is used
}

func TestUpdateStepReturnsMethod(t *testing.T) {
	// Verify the method signature exists by type assertion
	cfg := DefaultConfig()
	poller := NewAgentPoller(cfg)
	if poller == nil {
		t.Fatal("poller should not be nil")
	}
	// Method existence is verified at compile time
}

func TestInferTypeHint(t *testing.T) {
	tests := []struct {
		value    interface{}
		expected string
	}{
		{true, "Boolean"},
		{false, "Boolean"},
		{42, "Long"},
		{int32(42), "Long"},
		{int64(42), "Long"},
		{3.14, "Double"},
		{float32(3.14), "Double"},
		{"hello", "String"},
		{[]interface{}{1, 2, 3}, "List"},
		{map[string]interface{}{"a": 1}, "Map"},
		{struct{}{}, "Any"},
	}

	for _, tt := range tests {
		result := inferTypeHint(tt.value)
		if result != tt.expected {
			t.Errorf("inferTypeHint(%v) = %s, expected %s", tt.value, result, tt.expected)
		}
	}
}
